### ISSUE TITLE
Glob interval empty set is wrong.

### ISSUE THREAD

**Issue by linas** [opencog/atomspace#2502]
See discussion in #2498 -- the current glob-matching/set-intersection code is using `{0,0}` to represent "the empty set", but this is incorrect.  The actual fix is maybe a few lines of code in one or two places, and an appropriate unit test.

**Comment by sikefield3** [comment]
hi,   
 i'd like to do this, if it's ok...
i'm able to build and run it in docker.

**Comment by linas** [comment]
Hi,

I just spotted this ... sure, please! I guess you'd like some pointers for where to start? But if you think you can do it, then just create a pull request -- that's it...

**Comment by linas** [comment]
A place to start would be to read the problem description in #2498 and then look at the patch in #2498 and look at GlobInterval -- that is the thing that holds the lower and upper bounds  for matching. Hunt around for places where GlobInterval is used, make sure the bounds checks are reasonable.

**Comment by linas** [comment]
I mean - this is not glorious or exciting work, it just gets you familiar with the code base.

**Comment by sikefield3** [comment]
Cool!   
I'm going to have a look at the code in the next few days and then come back with my suggestion/further questions.

**Comment by sikefield3** [comment]
So, I did  a search for GlobInterval and now have some idea what to do...   
Imho, a function like `bool IsEmpty(GlobInterval)` would be handy, where the check according to 
[your suggestion](https://github.com/opencog/atomspace/pull/2498#issuecomment-585909835) is done.   

Anyway, I'm going to create my pull request soon...

**Comment by linas** [comment]
Yes, something like `IsEmpty()` sounds reasonable!

> I'm going to create my pull request soon...

Cool. FWIW, I sometimes don't check email for days in a row, don't panic if there's no immediate reply.

**Comment by sikefield3** [comment]
I'm getting an unresolved ref. for my function, so maybe I could make a WIP-pull-request and you could have a look, if I should change the function signature or location ?   
Apart from that the changes I wanted to make are in place.   
Unit tests can take while, since I'm not yet familiar with CxxTest.

**Comment by sikefield3** [comment]
[fix](https://github.com/opencog/atomspace/compare/master...sikefield3:atomspace:wip_is2502_new_empty_interval) is now done.   
unit tests still missing.

**Comment by sikefield3** [comment]
I was just wondering about the [extend](https://github.com/opencog/atomspace/blob/ef3c0a35b1814e20678d9f957d14d8b10669854d/opencog/atoms/core/TypeChoice.cc#L139-L146) -function:   
According to the comment, it returns the union of two intervals, but what if the intervals are separated: say `lhs = {0,2}, rhs = {4,6}`   ?
```
// Compute the union of two glob intervals.    
static inline GlobInterval extend(const GlobInterval& lhs,   
                                  const GlobInterval& rhs)   
{   
	const auto lb = std::min(lhs.first, rhs.first);   
	const auto ub = std::max(lhs.second, rhs.second);   
	return GlobInterval{lb, ub};   
}   
```

**Comment by linas** [comment]
> WIP

I'm looking now.

> Extend

Good question. I believe that an error should be thrown for such disjoint unions. There is a way of handling disjoint intervals, and I don't think it would flow through this code . Documented here:  https://wiki.opencog.org/w/GlobNode  bottom of page.  FWIW, the page also says " (This might not be implemented, yet. If it is, it might not be tested ... it might be broken!) "

The general guideline about such "fringe" cases is that if no one needs that capability, if no one is using it, it doesn't get implemented. i.e. that development is user-driven, demand-driven, so that we're not building castles that no one cares about. That's a "general guideline", though. If its easy to implement, it should be done.

**Comment by linas** [comment]
I merged the pull req. I guess this fixes everything?  Care to extend one of the unit tests to make sure?  The unit tests tend to have a nasty, ugly style, but still, give you some flavor of what's going on.

What's next after this? What do you want to do next?

**Comment by sikefield3** [comment]
> I merged the pull req. I guess this fixes everything? Care to extend one of the unit tests to make sure? The unit tests tend to have a nasty, ugly style, but still, give you some flavor of what's going on.
> 
Sure. I want to make some unit tests, but  I didn't figure them out quite yet...

**Comment by sikefield3** [comment]
> What's next after this? What do you want to do next?

how about throwing an exception for the disjoint intervals  (as a temporary  or permanent solution)?!

> Good question. I believe that an error should be thrown for such disjoint unions. There is a way of handling disjoint intervals, and I don't think it would flow through this code . Documented here: https://wiki.opencog.org/w/GlobNode bottom of page. FWIW, the page also says " (This might not be implemented, yet. If it is, it might not be tested ... it might be broken!) "

**Comment by linas** [comment]
> Sure. I want to make some unit tests, but I didn't figure them out quite yet...

There's an existing `GlobUTest` it can be espanded by emulating one of the other tests in there.

> how about throwing an exception for the disjoint intervals

Sure. That's not a bad way of signalling unimplemented code.

There are some meta-questions, perhaps the would be better discussed privately in email, but for now, I'll post here. Reply here or in private (or not at all  :-)

What do you hope to get out of this? There's more to the world than just the atomspace.  Let me ramble for a bit ---

The atomspace is mostly "done". Well, there's actually a lot of things that could be changed, improved and cleaned up, but exactly what and how remains unclear, because its not really use-case driven.  The biggest issue for the atomspace is that its too complicated for "ordinary programmers" to use on "ordinary projects", but I don't know how to solve this. It needs some kind of programming API guru.

Something Nii wanted was an atomese API for unification.  There's a block of code that does it, it's not very large, but it has an incredibly complex API. Figuring out how to make it easy is .. hard work. Its the stuff in the "unifier" directory of the URE. Medium priority. I don't need this, but URE-style thinkers do.

Support for frames needs to be implemented in the CogStorageNode, but this is lower-ish priority.

Port the RocksStorageNode to some other DB of your choice. Low priority but some people seem to be perenially interested in this stuff.

Mid-priority: build a plugin that would automatically map certain Atoms into certain live databases, and back. So, for example, given some DB containing genetic or biology data, say, having a table holding something, map each row of the table to an equivalent atomspace structure, "live" "dynamically", so that it pulls from the DB only when needed, and writes back to it only when required.  I think people might like this. It's a pretty big project, many months or half-a-year of work. It's got some complicated, abstract conceptual bit it make it challenging. ...

More AS-MOSES work...

I need help with assorted mundane stuff in the language-learning effort. This is the top-highest priority for me, taking up 98% of my attention span.  It would require learning and coding in scheme.  Although .. hmm ... some of the most "stable" and "well-understood" parts are ready to be converted to C++ for performance. These projects are varying sizes, many are "small" -- few days/weeks ...  For me, this is where the action is.

**Comment by sikefield3** [comment]
> There are some meta-questions, perhaps the would be better discussed privately in email, but for now, I'll post here. Reply here or in private (or not at all :-)

Ok! I'm going to send you an email !

**Comment by sikefield3** [comment]
Just spotted something:   
In [Variables::is_lower_bound](https://github.com/opencog/atomspace/blob/0d32e51d07ea2ffcd747c507f2f8187992ac702c/opencog/atoms/core/Variables.cc#L308) , shouldn't we return true, if `n == 0`, since this is the size difference between handle and pattern in `PatternMatchEngine::setup_rotors` (if I got this right) ?
(same with is_upper_bound.)

**Comment by linas** [comment]
> shouldn't we return true, if we have an empty interval  and ...

I suppose so. This is where unit tests start to really be important.  You want to imagine some situation where you have a glob, and it's OK for it to be empty. Then try it, and see if it works. If it works, great! If it doesn't, you've found a bug. The correct fix is then whatever actually fixes it. 

In this case,  lets see -- suppose you have data `(List A B C)` and you want to find all partitions. Then  match against `(List (Glob $X) (Glob $Y))` and the expected results are:
```
X=empty,   Y= A B C
X= A        Y= B C
X= A B      Y=C
X = A B C   Y = empty
```
Seems simple enough to me.   Either it works or it doesn't ... 

I do 100% of my scratch-pad testing by typing stuff in at the guile prompt. cut-n-paste till everything works. So the above, in glorious detail, would be
```
$ guile
scheme@(guile-user)>

(use-modules (opencog) (opencog exec))

(List (Concept "A") (Concept "B") (Concept "C"))
(define qry
   (Query
      (VariableList (Glob "$X") (Glob "$Y"))
      (List (Glob "$X") (Glob "$Y"))
      (OrderedLink (Concept "begin")  (Glob "$X")  (Concept "middle") (Glob "$Y") (Concept "end"))
  ))

(cog-execute! qry)
```

I just tried the above, it works, but I didn't try to allow empty Globs ... I'll leave that as homework for you.

To convert to a unit test, cut-n-paste into GlobUTest, or perhaps it will be easier to do what `or-bind.scm` does --- small, simple, easy-to-read.

**Comment by linas** [comment]
p.s. I'm impressed if you're looking at `PatternMatchEngine::setup_rotors` -- the code in `PatternMatchEngine` is the most complicated that there is in all of opencog.

**Comment by sikefield3** [comment]
> > shouldn't we return true, if we have an empty interval  and ...
> 
> I suppose so. This is where unit tests start to really be important. You want to imagine some situation where you have a glob, and it's OK for it to be empty. Then try it, and see if it works. If it works, great! If it doesn't, you've found a bug. The correct fix is then whatever actually fixes it.
> 
> In this case, lets see -- suppose you have data `(List A B C)` and you want to find all partitions. Then match against `(List (Glob $X) (Glob $Y))` and the expected results are:
> 
> ```
> X=empty,   Y= A B C
> X= A        Y= B C
> X= A B      Y=C
> X = A B C   Y = empty
> ```
> ...

I was playing around with this:    
The following query:
```
(define qry
   (Query
      (VariableList
		(TypedVariable (Glob "$X") (IntervalLink (Number 0) (Number -1)))
		(TypedVariable (Glob "$Y") (IntervalLink (Number 0) (Number -1))))
      (List (Glob "$X") (Glob "$Y"))
      (OrderedLink (Concept "begin") (Glob "$X")  (Concept "middle") (Glob "$Y") (Concept "end"))
  ))
```
gives the desired result.
, but if I set - say - 
```
(TypedVariable (Glob "$X") (IntervalLink (Number 1) (Number 0)))
```
the result is always empty (and not `X=empty,   Y= A B C` as I would expect). But with 
```
(TypedVariable (Glob "$X") (IntervalLink (Number 0) (Number 0)))
```
this works. 

> Just spotted something: In [Variables::is_lower_bound](https://github.com/opencog/atomspace/blob/0d32e51d07ea2ffcd747c507f2f8187992ac702c/opencog/atoms/core/Variables.cc#L308) , shouldn't we return true, if we have an empty interval and `n == 0`, since this is the size difference between handle and pattern in `PatternMatchEngine::setup_rotors` (if I got this right) ? (same with is_upper_bound.)

Changing Variables::is_lower_bound makes no difference.

**Comment by linas** [comment]
> I was playing around with this:
> `(TypedVariable (Glob "$X") (IntervalLink (Number 1) (Number 0)))`

Well, that sounds like a bug ... It depends on the meaning (semantics) of
`(IntervalLink (Number K) (Number N))`  whenever K>N.  There are two choices: either it means "the empty interval" or it means "no upper bound, the upper bound is infinite" Both seem to be reasonable defaults, both have a mathematical foundation (the second would be the so-called "one point compaction")  Until now, no one has had to answer this question ... I suppose you can flip a coin.

I suppose most ordinary programmers would be less surprised if the answer was "the empty interval" Probably the safest answer too ... because K and N can be computed.  For example:
```
(IntervalLink (Plus (Number 2) (number 2)) (Number 5))
```
but also wilder things:
```
(cog-set-value! (Concept "foo") (Predicate "bar") (FloatValue  2))
...
(IntervalLink 
    (Plus
        (Number 2)
        (FloatValueOf (Concept "foo") (Predicate "bar") ))
    (Number 5))
```
which, by the way, should work ...  The first part, you do manually: you `cog-set-value!` some value. The second part, whenever you run the query, it will automatically get the current value, add 2, and use that in the interval link.

Hmm... that should work. But as I think about it, I suspect that we don't ever actually execute variable declarations, so the Plus won't ever run.  So that's another "bug" that no one has probed before ...

**Comment by linas** [comment]
This *should* definitely work:
```
(cog-execute!
   (IntervalLink 
       (Plus
            (Number 2)
           (FloatValueOf (Concept "foo") (Predicate "bar")))
       (Number 5)))
```

I suspect it does not, because `IntervalLink` is not declared to be executable.  I don't know if this should be "fixed" or not. My general philosophy is "don't fix things until someone really needs it", which is balanced out by "Gee, this is so basic, it should just work without anyone asking for it."

**Comment by linas** [comment]
Closing. From what I can tell, all issues were fixed by @sikefield3  in pull reqs #3005 and #3017 (If not, if you disagree, please open a new issue and reference this one...)
